### PR TITLE
[UR][L0] Add UR_L0_LEAKS_DEBUG key

### PR DESCRIPTION
**Pull request by jandres742**
Use a new environment variable, UR_L0_LEAKS_DEBUG, to check for leaks in the UR L0 adapter, instead of relying on a specific value being set in UR_L0_DEBUG.

### CODE REVIEW

File: sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp
Comment (smaslov-intel) [inline sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: Why are we doing this explicitly in some (many) tests? I thought we run **all** tests twice, and one of the runs is with ZE_DEBUG=4, this way detecting leak in all tests.

Comment (jandres742) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: @smaslov-intel : I guess that is a question for the team running or designing the tests? Here I have changed only the references to ZE_DEBUG=4 to UR_L0_LEAKS_DEBUG=1, but whether they should be being used at all is something they are better positioned to answer. Let me know what you think.

Comment (smaslov-intel) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: Fair point, I will approve this change. But I'd like to hear from @intel/sycl-graphs-reviewers I guess about the motivation here.

Comment (smaslov-intel) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: Another point is that we should synchronize with how we run this testing for leaks, it should now use UR_L0_DEBUG instead of ZE_DEBUG.

Comment (jandres742) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: Agree, this will help us start moving to UR env vars.

Comment (EwanC) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: > I thought we run all tests twice, and one of the runs is with ZE_DEBUG=4, this way detecting leak in all tests.

We introduced this leak checking when we previously had a leak of L0 events in our UR command-buffer backend for graphs, and wanted to prevent this from regressing (or issues being introduced in new tests). Crucially, this was also at a time when we were developing on a fork of DPC++, so didn't have CI to catch non-default test configurations and wanted to catch the leaks during local development.

Could you point me to where we are running the tests a second time with `ZE_DEBUG=4` (or setting `ur_l0_leaks_debug` to pass to lit)? I think the suspicion that the graphs tests don't need their own extra leak checks is probably correct if that's the case.

Comment (jandres742) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:7]: thanks @EwanC .

@smaslov-intel : i think that question would be for you:

>Could you point me to where we are running the tests a second time with ZE_DEBUG=4 (or setting ur_l0_leaks_debug to pass to lit)? I think the suspicion that the graphs tests don't need their own extra leak checks is probably correct if that's the case.

**Comment by kbenzie** [comment]
I've merged https://github.com/oneapi-src/unified-runtime/pull/1053 now, due to timezones I'll update the UR tag in this PR myself so we can run testing earlier.

File: sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp
Comment (EwanC) [reply sycl/test-e2e/Graph/Explicit/add_node_while_recording.cpp:6]: To help unblock this discussion, if you know where this second run is happening @smaslov-intel,  then create a GitHub Issue with that information and assign it to me, and I'll make this change to the graphs to remove the leak checking run from the graphs E2E tests.

**Review by aarongreig — APPROVED** [review]
UR LGTM

**Comment by kbenzie** [comment]
The graph test failures [here](https://github.com/intel/llvm/actions/runs/6956750746/job/18929142684?pr=11820) were added recently so likely need updated in the same way the other tests have been in this PR. I'll try reproducing locally and applying the fix here if I'm able to.

**Comment by kbenzie** [comment]
@intel/llvm-reviewers-runtime please review

**Review by jandres742 — COMMENTED** [review]
+1

**Comment by jandres742** [comment]
> The graph test failures [here](https://github.com/intel/llvm/actions/runs/6956750746/job/18929142684?pr=11820) were added recently so likely need updated in the same way the other tests have been in this PR. I'll try reproducing locally and applying the fix here if I'm able to.

thanks @kbenzie !

File: sycl/test-e2e/Plugin/interop-level-zero-buffer.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/Plugin/interop-level-zero-buffer.cpp:4]: This and same in other files needs to be removed (or rather changed to ur_l0_debug) as part of this PR.

File: sycl/test-e2e/lit.cfg.py
Comment (aelovikov-intel) [inline sycl/test-e2e/lit.cfg.py:141]: I don't think we need this, because it would result in the same issue we're trying to fix.

**Review by aelovikov-intel — COMMENTED** [review]
I think proper separation would be `UR_L0_TRACE` vs `UR_L0_VERIFY` so that validation layer could be enabled in the latter instead of `UR_L0_DEBUG` which would have the same issues as before this PR. I won't insist on that part though because I don't remember a single SYCL RT issue caught by such validation and I can live without it in our testing.

File: sycl/test-e2e/lit.cfg.py
Comment (jandres742) [reply sycl/test-e2e/lit.cfg.py:141]: thansk @aelovikov-intel . I'm not familiar with this code, so I will remove it if wanted.

@smaslov-intel , @EwanC : what do you think?

File: sycl/test-e2e/Plugin/interop-level-zero-buffer.cpp
Comment (jandres742) [reply sycl/test-e2e/Plugin/interop-level-zero-buffer.cpp:4]: I can remove that.

**Comment by jandres742** [comment]
> I think proper separation would be `UR_L0_TRACE` vs `UR_L0_VERIFY` so that validation layer could be enabled in the latter instead of `UR_L0_DEBUG` which would have the same issues as before this PR. I won't insist on that part though because I don't remember a single SYCL RT issue caught by such validation and I can live without it in our testing.

that's a good idea. I think we could do that in a follow-up PR to further separate the functionality of UR_L0_DEBUG.

**Review by smaslov-intel — APPROVED** [review]

File: sycl/test-e2e/lit.cfg.py
Comment (smaslov-intel) [reply sycl/test-e2e/lit.cfg.py:141]: I think it is good to have the ability to request debug traces with an option

Comment (EwanC) [reply sycl/test-e2e/lit.cfg.py:141]: I don't really have an opinion on this. Presumably a user could pass `lit --param extra_environment="UR_L0_DEBUG=1"` to request debug traces regardless, but I don't object to having a convenience option.

**Comment by kbenzie** [comment]
Seems like some changes are needed on this PR. I'd really appriciate if we could get this in a ready to marge state ASAP though as this is currently a blocker for https://github.com/intel/llvm/pull/11718 and then https://github.com/intel/llvm/pull/11893 which are also both intended for the next release.

**Comment by jandres742** [comment]
@againull, @intel/llvm-gatekeepers : could you help us with the approval and merge of this PR? We cannot bring any change from UR until this is merged.

**Comment by steffenlarsen** [comment]
@aelovikov-intel has open topics on this PR. I will let him decide if his concerns have been addressed.

**Comment by kbenzie** [comment]
> @aelovikov-intel has open topics on this PR. I will let him decide if his concerns have been addressed.

If there are unresolved things could we create and assign issues to be followed up on later we can get UR PR's moving again?

**Comment by jandres742** [comment]
thanks @steffenlarsen ! Correct, those comments can be done on follow-ups, as @kbenzie indicated.

**Pull request merged**
